object_intern.go

```diff
@@ -100,49 +103,88 @@ func (oi *ObjectIntern) DecompressString(in string) (string, error) {
 // If the object is found in the store its reference count is increased by 1.
 // If the object is added to the store its reference count is set to 1.
 func (oi *ObjectIntern) AddOrGet(obj []byte, safe bool) (uintptr, error) {
-	var addr gos.ObjAddr
-	var ok bool
-	var err error
-	var objString string
 
-	objComp := obj
+	// if either of these two terms is true then the rest of this block
+	// requires a lot of allocations
+	if (oi.conf.Compression != None) || (safe && oi.conf.Compression == None) {
+		objComp := obj
 
-	if oi.conf.Compression != None {
-		objComp = oi.compress(obj)
-	}
+		if oi.conf.Compression != None {
+			// this returns a new byte slice, so we don't need to check for safe
+			objComp = oi.compress(obj)
+		}
 
-	// the only case we need to handle specially is when compression
-	// is turned off and the user has requested that the operation be safe
-	if safe && oi.conf.Compression == None {
-		// create a copy so we don't modify the original []byte
-		objComp = make([]byte, len(obj))
-		copy(objComp, obj)
-	}
+		// the only case we need to handle specially is when compression
+		// is turned off and the user has requested that the operation be safe
+		if safe && oi.conf.Compression == None {
+			// create a copy so we don't modify the original []byte
+			objComp = make([]byte, len(obj))
+			copy(objComp, obj)
+		}
 
-	// string used for the index
-	objString = string(objComp)
+		// acquire lock
+		oi.Lock()
+
+		// try to find the object in the index
+		addr, ok := oi.objIndex[string(objComp)]
```

Comment:
just a random idea: would this index lookup over millions of entries potentially be faster if `oi.objIndex` is sharded by object length? like `oi.objIndex[uint16][string]` instead of `oi.objIndex[string]`

Reply:
Unrelated to the above comment:
Now `obj` gets copied into `objComp` and then `objComp` is used to check whether this object is already present. I assume in the majority of cases it is already present, in which case the copying was an unnecessary allocation. Might be worth it to first use `obj` just to check whether the object is there, and only if it isn't copy & insert it

Reply:
In regards to your first comment. I did give that a lot of thought, especially with lock sharding. It's similar to what I was looking at for the other index we are working on. It probably would speed things up when we have a lot of entries, but with the nature of this data we will still have a lot of hotspots. Specifically between lengths ~3-10. My main concern is that the goal of all of this is to reduce memory consumption, so all of my decisions are solely focused on that right now. If speed / contention becomes an issue in the future I would love to explore different ways of sharding the data / locks.

As to your second (unrelated) comment, it's a little tricky. If compression is turned on then we must compress it, so the allocation will happen regardless. If compression is not turned on, but the user has specified that they would like the operation to be safe, then we are going to make a copy of it. In this case we could avoid the allocation if the object already exists in the store. Chances are very high that after a program has been running and using the interning layer for even a little bit of time that all of the interned objects will already be present in the store.

I will look at changing it so that if compression is turned off and they want the operation to be safe that it doesn't allocate until after the check.

Reply:
Sharded index will have to wait until I get some performance stats to see how it does at a larger scale.

I implemented the pre-allocation checks.